File: js/boj/9252.js
/*
[boj] LCS2 (9252)
https://www.acmicpc.net/problem/9252
*/

// 파일로부터 값 입력 받기
const fs = require("fs");
const filePath =
	process.platform === "linux" ? "/dev/stdin" : "./testcase/9252.txt";
let input = fs.readFileSync(filePath).toString().trim().split("\n");
const s1 = " " + input[0];
const s2 = " " + input[1];

let dy = [...Array(s1.length)].map((e) => Array(s2.length).fill(""));

for (let i = 1; i < s1.length; i++) {
	for (let j = 1; j < s2.length; j++) {
		if (s1[i] === s2[j]) {
			dy[i][j] = dy[i - 1][j - 1] + s1[i];
		} else {
			if (dy[i][j - 1].length >= dy[i - 1][j].length)
				dy[i][j] = dy[i][j - 1];
			else dy[i][j] = dy[i - 1][j];
		}
	}
}
console.log(dy[s1.length - 1][s2.length - 1].length);
console.log(dy[s1.length - 1][s2.length - 1]);

File: c++/boj/9252.cpp
/*
 * LCS 2
 * https://www.acmicpc.net/problem/9252
 */

#include <bits/stdc++.h>
using namespace std;

int main() {
	string s1;	cin >> s1;	s1 = " " + s1;
	string s2;	cin >> s2;	s2 = " " + s2;
	vector<vector<string> > dp(s1.length(), vector<string>(s2.length()));
	for (int i = 1; i < s1.length(); i++) {
		for (int j = 1; j < s2.length(); j++) {
			if (s1[i] == s2[j])
				dp[i][j] = dp[i - 1][j - 1] + s1[i];
			else if (dp[i][j - 1].length() >= dp[i - 1][j].length())
				dp[i][j] = dp[i][j - 1];
			else
				dp[i][j] = dp[i - 1][j];
		}
	}
	cout << dp[s1.length() - 1][s2.length() - 1].length() << '\n';
	cout << dp[s1.length() - 1][s2.length() - 1];
}